File: examples/ocaml/x20.ml
(*
  plimage demo

*)

open Plplot

let pi = atan 1.0 *. 4.0

let xdim = 260
let ydim = 220

(* Change this to 'Some "outfile.psc"' if you want to test the plot saving
   option. *)
let filename = None

type stretch_data = {
  xmin : float;
  xmax : float;
  ymin : float;
  ymax : float;
  stretch : float;
}

(* Transformation function *)
let mypltr pltr_data x y =
  let x0 = (pltr_data.xmin +. pltr_data.xmax) *. 0.5 in
  let y0 = (pltr_data.ymin +. pltr_data.ymax) *. 0.5 in
  let dy = (pltr_data.ymax -. pltr_data.ymin) *. 0.5 in
  x0 +. (x0 -. x) *.
    (1.0 -. pltr_data.stretch *. cos ((y -. y0) /. dy *. pi *. 0.5)),
  y

(* read image from file in binary ppm format *)
let read_img fname =
  (* naive grayscale binary ppm reading. If you know how to, improve it *)
  let fp = open_in fname in

  let ver = input_line fp in (* version *)

  if ver <> "P5" then
    raise (Invalid_argument ("Unable to read image: " ^ fname))
  else
    ();

  (* This will skip comments. *)
  let rec ignore_comments_then_get_w_h () =
    let maybe_comment = input_line fp in
    if maybe_comment.[0] = '#' then
      ignore_comments_then_get_w_h ()
    else
      (* This line has the width and height in it *)
      maybe_comment
  in

  (* width, height num colors *)
  let w_h_line = ignore_comments_then_get_w_h () in
  let num_col_line = input_line fp in
  let w, h = Scanf.sscanf w_h_line "%d %d" (fun w h -> w, h) in
  let num_col = Scanf.sscanf num_col_line "%d" (fun n -> n) in

  let img = String.make (w * h) ' ' in
  let imf = Array.make_matrix w h 0.0 in

  (* Note that under 32bit OCaml, this will only work when reading strings up
     to ~16 megabytes. *)
  really_input fp img 0 (w * h);

  close_in fp;

  for i = 0 to w - 1 do
    for j = 0 to h - 1 do
      imf.(i).(j) <-
        (* flip image up-down *)
        float_of_int (int_of_char (img.[(h - 1 - j ) * w + i]));
    done
  done;
  imf, w, h, num_col

(* save plot *)
let save_plot fname =
  let cur_strm = plgstrm () in (* get current stream *)
  let _ = plmkstrm () in  (* create a new one *)

  plsdev "psc"; (* new device type. Use a known existing driver *)
  plsfnam fname; (* file name *)

  plcpstrm cur_strm false; (* copy old stream parameters to new stream *)
  plreplot (); (* do the save *)
  plend1 (); (* close new device *)

  plsstrm cur_strm; (* and return to previous one *)
  ()

(* set gray colormap *)
let gray_cmap num_col =
  let r = [|0.0; 1.0|] in
  let g = [|0.0; 1.0|] in
  let b = [|0.0; 1.0|] in
  let pos = [|0.0; 1.0|] in
  plscmap1n num_col;
  plscmap1l true pos r g b None;
  ()

let () =
  (* Parse and process command line arguments *)
  plparseopts Sys.argv [PL_PARSE_FULL];

  (* Initialize plplot *)
  plinit ();

  let z = Array.make_matrix xdim ydim 0.0 in
  let r = Array.make_matrix xdim ydim 0.0 in

  (* sombrero-like demo *)
  plcol0 2; (* draw a yellow plot box, useful for diagnostics *)
  plenv 0.0 (2.0 *. pi) 0.0 (3.0 *. pi) 1 (-1);

  let x =
    Array.init xdim (
      fun i -> float_of_int i *. 2.0 *. pi /. float_of_int (xdim - 1)
    )
  in
  let y =
    Array.init ydim (
      fun i -> float_of_int i *. 3.0 *. pi /. float_of_int (ydim - 1)
    )
  in

  for i = 0 to xdim - 1 do
    for j = 0 to ydim - 1 do
      r.(i).(j) <- sqrt (x.(i) *. x.(i) +. y.(j) *. y.(j)) +. 1e-3;
      z.(i).(j) <- sin r.(i).(j) /. r.(i).(j);
    done
  done;

  pllab "No, an amplitude clipped \"sombrero\"" "" "Saturn?";
  plptex 2.0 2.0 3.0 4.0 0.0 "Transparent image";
  plimage
    z 0.0 (2.0 *. pi) 0.0 (3.0 *. pi) 0.05 1.0 0.0 (2.0 *. pi) 0.0 (3.0 *. pi);

  (* save the plot *)
  let () =
    match filename with
        Some f -> save_plot f
      | None -> ()
  in

  (* read Chloe image *)
  (* Note we try two different locations to cover the case where this
   * examples is being run from the test_ocaml.sh script *)
  let img_f, width, height, num_col =
    try
      read_img "Chloe.pgm"
    with
      _ -> (
        try
          read_img "../Chloe.pgm"
        with
          Invalid_argument _ ->
            plend ();
            raise (Failure "No such file")
      )
  in

  let width = float_of_int width in
  let height = float_of_int height in

  (* set gray colormap *)
  gray_cmap num_col;

  (* display Chloe *)
  plenv 1.0 width 1.0 height 1 (-1);

  pllab "Set and drag Button 1 to (re)set selection, Button 2 to finish." " " "Chloe...";

  plimage img_f 1.0 width 1.0 height 0.0 0.0 1.0 width 1.0 height;

  (* selection/expansion demo *)
  let xi = 25.0 in
  let xe = 130.0 in
  let yi = 235.0 in
  let ye = 125.0 in

  plspause false;
  pladv 0;

  (* display selection only *)
  plimage img_f 1.0 width 1.0 height 0.0 0.0 xi xe ye yi;

  plspause true;

  (* zoom in selection *)
  plenv xi xe ye yi 1 (-1);
  plimage img_f 1.0 width 1.0 height 0.0 0.0 xi xe ye yi;

  (* Base the dynamic range on the image contents. *)
  let img_max, img_min = plMinMax2dGrid img_f in

  (* Draw a saturated version of the original image.  Only use the middle 50%
     of the image's full dynamic range. *)
  plcol0 2;
  plenv 0.0 width 0.0 height 1 (-1);
  pllab "" "" "Reduced dynamic range image example";
  plimagefr
    img_f 0.0 width 0.0 height 0.0 0.0
    (img_min +. img_max *. 0.25) (img_max -. img_max *. 0.25);

  (* Draw a distorted version of the original image, showing its full dynamic
     range. *)
  plenv 0.0 width 0.0 height 1 (-1);
  pllab "" "" "Distorted image example";

  let stretch =
    {
      xmin = 0.0;
      xmax = width;
      ymin = 0.0;
      ymax = height;
      stretch = 0.5;
    }
  in

  (* OCaml supports arbitrary coordinate transform callbacks, so take advantage
     of that to use mypltr directly and skip the use of pltr2. *)
  plset_pltr (mypltr stretch);
  (* Comment the above line, and uncomment the following section to use pltr2 *)
  (*
  let xg = Array.make_matrix (int_of_float width + 1) (int_of_float height + 1) 0.0 in
  let yg = Array.make_matrix (int_of_float width + 1) (int_of_float height + 1) 0.0 in
  for i = 0 to int_of_float width do
    for j = 0 to int_of_float height do
      let xx, yy = mypltr (float_of_int i) (float_of_int j) stretch in
      xg.(i).(j) <- xx;
      yg.(i).(j) <- yy;
    done
  done;
  plset_pltr (pltr2 xg yg);
  *)
  plimagefr img_f 0.0 width 0.0 height 0.0 0.0 img_min img_max;

  plend ();
  ()

File: examples/lua/x20.lua
--[[
	plimage demo
--]]

-- initialise Lua bindings for PLplot examples.
dofile("plplot_examples.lua")

XDIM = 260
YDIM = 220

dbg = 0
nosombrero = 0
nointeractive = 0
f_name=""


-- Transformation function
function mypltr(x,  y)
  local x0 = (stretch["xmin"] + stretch["xmax"])*0.5
  local y0 = (stretch["ymin"] + stretch["ymax"])*0.5
  local dy = (stretch["ymax"]-stretch["ymin"])*0.5
  local tx = x0 + (x0-x)*(1 - stretch["stretch"]*math.cos((y-y0)/dy*math.pi*0.5))
  local ty = y

  return tx, ty
end


-- read image from file in binary ppm format
function read_img(fname)
  -- naive grayscale binary ppm reading. If you know how to, improve it
  local fp = io.open(fname, "rb")
  if fp==nil then
    return 1
  end

  -- version
  local ver = fp:read("*line")

  if ver~="P5" then -- I only understand this!
    fp:close()
    return 1
  end

  while fp:read(1)=="#" do
    local com = fp:read("*line")
    if com==nil then
      fp:close()
      return 1
    end
  end
  fp:seek("cur", -1)

  local w, h, num_col = fp:read("*number", "*number", "*number")
  if w==nil or h==nil or num_col==nil then -- width, height, num colors
    fp:close()
    return 1
  end

  -- read the rest of the line (only EOL)
  fp:read("*line")

  local img = fp:read(w*h)
  fp:close()
  if string.len(img)~=(w*h) then
    return 1
  end

  local imf = {}

  for i = 1, w do
    imf[i] = {}
    for j = 1, h do
      imf[i][j] = string.byte(img, (h-j)*w+i) -- flip image up-down
    end
  end

  return 0, imf, w, h, num_col
end


-- save plot
function save_plot(fname)
  local cur_strm = pl.gstrm() -- get current stream
  local new_strm = pl.mkstrm() -- create a new one

  pl.sdev("psc") -- new device type. Use a known existing driver
  pl.sfnam(fname) -- file name

  pl.cpstrm(cur_strm, 0) -- copy old stream parameters to new stream
  pl.replot()	-- do the save
  pl.end1() -- close new device

  pl.sstrm(cur_strm)	-- and return to previous one
end


--  get selection square interactively
function get_clip(xi, xe, yi, ye)
  return 0, xi, xe, yi, ye
end


-- set gray colormap
function gray_cmap(num_col)
  local r = { 0, 1 }
  local g = { 0, 1 }
  local b = { 0, 1 }
  local pos = { 0, 1 }

  pl.scmap1n(num_col)
  pl.scmap1l(1, pos, r, g, b)
end


x = {}
y = {}
z = {}
r = {}
img_f = {}
cgrid2 = {}


-- Parse and process command line arguments
pl.parseopts(arg, pl.PL_PARSE_FULL)

-- Initialize plplot
pl.init()

z={}

-- view image border pixels
if dbg~=0 then
  pl.env(1, XDIM, 1, YDIM, 1, 1) -- no plot box

  -- build a one pixel square border, for diagnostics
  for i = 1, XDIM do
    z[i] = {}
    z[i][1] = 1 -- left
    z[i][YDIM] = 1 -- right
  end

  for i = 1, YDIM do
    z[1][i] = 1 -- top
    z[XDIM][i] = 1 -- botton
  end

  pl.lab("...around a blue square."," ","A red border should appear...")

  pl.image(z, 1, XDIM, 1, YDIM, 0, 0, 1, XDIM, 1, YDIM)
end

-- sombrero-like demo
if nosombrero==0 then
  r = {}
  pl.col0(2) -- draw a yellow plot box, useful for diagnostics! :(
  pl.env(0, 2*math.pi, 0, 3*math.pi, 1, -1)

  for i = 1, XDIM do
    x[i] = (i-1)*2*math.pi/(XDIM-1)
  end
  for i = 1, YDIM do
    y[i] = (i-1)*3*math.pi/(YDIM-1)
  end

  for i = 1, XDIM do
    r[i] = {}
    z[i] = {}
    for j=1, YDIM do
      r[i][j] = math.sqrt(x[i]^2+y[j]^2)+1e-3
      z[i][j] = math.sin(r[i][j])/r[i][j]
    end
  end

  pl.lab("No, an amplitude clipped \"sombrero\"", "", "Saturn?")
  pl.ptex(2, 2, 3, 4, 0, "Transparent image")
  pl.image(z, 0, 2*math.pi, 0, 3*math.pi, 0.05, 1, 0, 2*math.pi, 0, 3*math.pi)

  -- save the plot
  if f_name~="" then
    save_plot(f_name)
  end
end

-- read Chloe image
-- Note we try two different locations to cover the case where this
-- examples is being run from the test_c.sh script
status, img_f, width, height, num_col = read_img("Chloe.pgm")
if status~=0 then
  status, img_f, width, height, num_col = read_img("../Chloe.pgm")
  if status~=0 then
    pl.abort("No such file")
    pl.plend()
    os.exit()
  end
end

-- set gray colormap
gray_cmap(num_col)

-- display Chloe
pl.env(1, width, 1, height, 1, -1)

if nointeractive==0 then
  pl.lab("Set and drag Button 1 to (re)set selection, Button 2 to finish."," ","Chloe...")
else
  pl.lab(""," ","Chloe...")
end

pl.image(img_f, 1, width, 1, height, 0, 0, 1, width, 1, height)

-- selection/expansion demo
if nointeractive==0 then
  xi = 25
  xe = 130
  yi = 235
  ye = 125

  status, xi, xe, yi, ye = get_clip(xi, xe, yi, ye)
  if status~=0 then  -- get selection rectangle
    pl.plend()
    os.exit()
  end

  pl.spause(0)
  pl.adv(0)

  -- display selection only
  pl.image(img_f, 1, width, 1, height, 0, 0, xi, xe, ye, yi)

  pl.spause(1)

  -- zoom in selection
  pl.env(xi, xe, ye, yi, 1, -1)
  pl.image(img_f, 1, width, 1, height, 0, 0, xi, xe, ye, yi)
end

-- Base the dynamic range on the image contents.
img_max, img_min = pl.MinMax2dGrid(img_f)

-- Draw a saturated version of the original image.  Only use the middle 50%
-- of the image's full dynamic range.
pl.col0(2)
pl.env(0, width, 0, height, 1, -1)
pl.lab("", "", "Reduced dynamic range image example")
pl.imagefr(img_f, 0, width, 0, height, 0, 0, img_min + img_max*0.25, img_max - img_max*0.25)

-- Draw a distorted version of the original image, showing its full dynamic range.
pl.env(0, width, 0, height, 1, -1)
pl.lab("", "", "Distorted image example")

stretch = {}
stretch["xmin"] = 0
stretch["xmax"] = width
stretch["ymin"] = 0
stretch["ymax"] = height
stretch["stretch"] = 0.5

-- In C / C++ the following would work, with plimagefr directly calling
-- mypltr. For compatibilty with other language bindings the same effect
-- can be achieved by generating the transformed grid first and then
-- using pltr2.
-- pl.imagefr(img_f, width, height, 0., width, 0., height, 0., 0., img_min, img_max, mypltr, (PLPointer) &stretch)

cgrid2 = {}
cgrid2["xg"] = {}
cgrid2["yg"] = {}
cgrid2["nx"] = width+1
cgrid2["ny"] = height+1

for i = 1, width+1 do
  cgrid2["xg"][i] = {}
  cgrid2["yg"][i] = {}
  for j = 1, height+1 do
    xx, yy = mypltr(i, j)
    cgrid2["xg"][i][j] = xx
    cgrid2["yg"][i][j] = yy
  end
end

--pl.imagefr(img_f, 0, width, 0, height, 0, 0, img_min, img_max, "pltr2", cgrid2)
pl.imagefr(img_f, 0, width, 0, height, 0, 0, img_min, img_max, "mypltr")

pl.plend()
